// node-red-contrib-controltimer/dist/node-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.constants = exports.defaults = exports.nodeName = void 0;
const timer_1 = require("./timer");
exports.nodeName = 'controltimer';
exports.defaults = {
    name: '',
    timerType: timer_1.TIMER_TYPE.DELAY,
    timerDurationUnit: timer_1.DurationUnit.SECOND,
    timerDurationType: 'num',
    timerDuration: 5,
    timerLoopTimeoutUnit: timer_1.DurationUnit.SECOND,
    timerLoopTimeoutType: 'num',
    timerLoopTimeout: 0,
    loopTimeoutMessageType: 'str',
    loopTimeoutMessage: 'LOOP_TIMEOUT',
    timerMaxLoopIterationsType: 'num',
    timerMaxLoopIterations: 0,
    loopMaxIterationsMessageType: 'str',
    loopMaxIterationsMessage: 'MAX_LOOP_ITERATIONS',
    isRunningTimerProgressVisible: true,
    outputReceivedMessageOnTimerTrigger: true,
    outputReceivedMessageOnTimerHalt: false,
    startTimerOnReceivalOfUnknownMessage: false,
    resetTimerOnReceivalOfUnknownMessage: false,
    resetTimerOnReceivalOfStartAction: false,
    continueTimerOnReceivalOfStartAction: false,
    isDebugModeEnabled: false,
    timerTriggeredMessageType: 'str',
    timerTriggeredMessage: 'TIMER_TRIGGERED',
    timerHaltedMessageType: 'str',
    timerHaltedMessage: 'TIMER_HALTED',
    isStartActionEnabled: true,
    isStopActionEnabled: true,
    isResetActionEnabled: true,
    isPauseActionEnabled: true,
    isContinueActionEnabled: true,
    actionPropertyNameType: 'str',
    actionPropertyName: 'payload',
    startActionNameType: 'str',
    startActionName: 'START',
    stopActionNameType: 'str',
    stopActionName: 'STOP',
    resetActionNameType: 'str',
    resetActionName: 'RESET',
    pauseActionNameType: 'str',
    pauseActionName: 'PAUSE',
    continueActionNameType: 'str',
    continueActionName: 'CONTINUE',
};
exports.constants = {
    timerTypeOverridePropertyName: 'timerType',
    timerDurationOverridePropertyName: 'timerDuration',
    timerDurationUnitOverridePropertyName: 'timerDurationUnit',
};
